File: Greedy/paintHouse/paintHouse.js
(function (){



    function minCost(arr,n,d) {
        let currentCost = arr[0].cost;
        let currentWorker = 0;
        let areaCoverd = 0;

        for(let i =1; i<n; i++) {
            areaCoverd += (arr[i].time - arr[i-1].time) * (arr[currentWorker].speed);
            if (areaCoverd >= d) return curruentCost;

            if (arr[currentWorker].speed < arr[i].speed) {
                currentWorker = i;
                currentCost += arr[i].cost;
            }
        }

        return currentCost;
    }

    (function main(){


        let totalWorkers = 3;
        let totalHouse = 3;
        let workers = [];
        let time = [1,1,1];
        let cost = [2,2,2];
        let speed = [3,1,5];

        for(let i=0; i<totalWorkers; i++) {
            let worker = {};
            worker.time = time[i];
            worker.cost = cost[i];
            worker.speed = speed[i];
            workers.push(worker);
        }

        workers.sort((a,b) => {
            if (a.time === b.time) {
                if (a.speed === b.speed) {
                    if (a.cost <= b.cost) {
                        return -1;
                    } else {
                        return 1;
                    }
                } else if (a.speed > b.speed){
                    return -1;
                } else {
                    return 1;
                }
            } else if(a.time < b.time) {
                return -1;
            } else {
                return 1;
            }
        });

        console.log(minCost(workers,workers.length,totalHouse));

    }());




}());

File: Greedy/paintHouse/paintHouse.cpp
#include<iostream>
#include<algorithm>
using std::cin;
using std::cout;
using std::endl;

struct Workers {
    int time;
    int cost;
    int speed;
};

bool compare(Workers a, Workers b) {
    // If two workers available at same time
    if (a.time == b.time) {
        // then choose one which have maximum speed
        if (a.speed == b.speed) {
            // if speed is same then go with whose cost is less, because ultimatly we need to maintain minimum cost
            return a.cost < b.cost;
        }
        return a.speed > b.speed;
    }
    return a.time < b.time;
}

long long int minCost(Workers* arr, int n, long long d) {
    long long currentCost {arr[0].cost};
    long long areaCoverd {};
    int currentWorker {};

    for(int i{1}; i < n && areaCoverd < d; i++) {
        areaCoverd = areaCoverd + (arr[i].time - arr[i-1].time) * arr[currentWorker].speed;

        if(areaCoverd >= d) return currentCost;

        if (arr[currentWorker].speed < arr[i].speed) {
            currentWorker = i;
            currentCost += arr[currentWorker].cost;
        }
    }
    return currentCost;
}

int main() {
    int n;
    cin >> n;

    long long d;
    cin >> d;

    Workers* arr{new Workers[n]};
    for(int i{}; i < n; i++) cin >> arr[i].time >> arr[i].cost >> arr[i].speed;

    long long ans{minCost(arr, n, d)};
    cout << ans << endl;
}
